src/order/countable_dense_linear_order.lean

```diff
@@ -0,0 +1,288 @@
+
+@[simp] lemma cmp_eq_gt_iff : cmp x y = ordering.gt ↔ y < x :=
+ordering.compares.eq_gt $ cmp_compares x y
+
+@[simp] lemma cmp_self_eq_eq : cmp x x = ordering.eq :=
+(ordering.compares.eq_eq (cmp_compares x x)).mpr rfl
+
+variables {x y} {β : Type*} [decidable_linear_order β] {x' y' : β}
+
+lemma cmp_eq_cmp_symm : cmp x y = cmp x' y' ↔ cmp y x = cmp y' x' :=
+by { split, rw [←cmp_swap _ y, ←cmp_swap _ y'], cc,
+  rw [←cmp_swap _ x, ←cmp_swap _ x'], cc, }
+
+lemma lt_iff_lt_of_cmp_eq (h : cmp x y = cmp x' y') : x < y ↔ x' < y' :=
+by rw [←cmp_eq_lt_iff, ←cmp_eq_lt_iff, h]
+
+lemma gt_iff_gt_of_cmp_eq (h : cmp x y = cmp x' y') : x ≤ y ↔ x' ≤ y' :=
+begin
+  suffices : ¬ y < x ↔ ¬ y' < x',
+  { push_neg at this, assumption },
+  rw not_iff_not,
+  apply lt_iff_lt_of_cmp_eq,
+  rwa cmp_eq_cmp_symm,
+end
+
+end cmp
+
+section orderiso
+
+variables {α : Type*} {β : Type*} [decidable_linear_order α] [decidable_linear_order β]
+  (f : α → β) (g : β → α)
+
+/-- To construct an order isomorphism between decidable linear orders `α` and `β`, it
+  suffices to construct maps `f : α → β`, `g : β → α`, satisfying
+  `∀ (a : α) (b : β), cmp a (g b) = cmp (f a) b`. We think of this as a two-sided Galois
+  adjunction. -/
+def order_iso_of_cmp_eq_cmp (adj : ∀ (a : α) (b : β), cmp a (g b) = cmp (f a) b) :
+  ((≤) : α → α → Prop) ≃o ((≤) : β → β → Prop) :=
+{ to_fun := f,
+  inv_fun := g,
+  left_inv := begin
+    intro a,
+    suffices : cmp a (g $ f a) = ordering.eq,
+    { symmetry, simpa using this },
+    rw adj, simp,
+  end,
+  right_inv := begin
+    intro b,
+    suffices : cmp (f $ g b) b = ordering.eq,
+    { simpa using this },
+    rw ←adj, simp,
+  end,
+  ord' := begin
+    intros x y,
+    apply gt_iff_gt_of_cmp_eq,
+    suffices : y = g (f y),
+    { convert adj x (f y), },
+    suffices : cmp y (g $ f y) = ordering.eq,
+    { simpa using this },
+    rw adj, simp,
+  end }
+
+end orderiso
+
+section swap
+
+lemma swap_eq_iff {α β} {p : α × β} {q : β × α} : p.swap = q ↔ p = q.swap :=
+begin
+  split,
+  { intro h, rw [←h, prod.swap_swap] },
+  { intro h, rw [h, prod.swap_swap] }
+end
+
+lemma mem_swap {α β} [decidable_eq α] [decidable_eq β] {s : finset $ α × β} {p : β × α} :
+  p ∈ s.image prod.swap ↔ p.swap ∈ s :=
+begin
+  rw finset.mem_image,
+  exact ⟨λ ⟨q, h, he⟩, swap_eq_iff.mp he ▸ h, λ h, ⟨p.swap, h, prod.swap_swap p⟩⟩,
+end
+
+lemma sub_swap_iff {α β} [decidable_eq α] [decidable_eq β] {s : finset $ α × β}
+  {t : finset $ β × α} : s ⊆ t.image prod.swap ↔ s.image prod.swap ⊆ t :=
+begin
+  rw [finset.subset_iff, finset.subset_iff],
+  split,
+  { intros h p hp,
+    rw mem_swap at hp,
+    specialize h hp,
+    rwa [mem_swap, prod.swap_swap] at h, },
+  { intros h p hp,
+    rw mem_swap,
+    apply h,
+    rwa [mem_swap, prod.swap_swap], }
+end
+
+end swap
+
+-- For any finsets lo, hi, with all elements of lo less than those of hi,
+-- there is an element m that separates them.
+lemma separation {α : Type*} [decidable_linear_order α]
```

Comment:
Since we're classical and noncomputable anyway, might as well use `linear_order` instead of `decidable_linear_order` here, I think